ENH: integrate: add args argument to solve_ivp.

WarrenWeckesser opened the pull request:
This pull request adds`args` to `scipy.integrate.solve_ivp`.  See the discussion in https://github.com/scipy/scipy/issues/8352.

Here's a demonstration script, "rossler_demo.py", that exercises the callable arguments `fun`, `jac` and `events`.

```

import numpy as np
from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt


def rossler_sys(t, w, a, b, c):
    """
    Rössler system
    """
    x, y, z = w
    return [-y - z, x + a*y, b +z*(x - c)]


def rossler_jac(t, w, a, b, c):
    """
    Jacobian of the Rössler system.
    """
    x, y, z = w
    dxs = [0, -1, -1]
    dys = [1, a, 0]
    dzs = [z, 0, x - c]
    return [dxs, dys, dzs]


def x0event(t, w, a, b, c):
    """
    x = 0 event for the Rössler system.
    """
    x, y, z = w
    return x

x0event.direction = 1


# Rössler parameters
a = 0.1
b = 0.1
c = 14.0

w0 = [1, 1, 1]

print("Computing transient")
sol_trans = solve_ivp(rossler_sys, [0, 100], w0, args=(a, b, c))

print("Computing long solution, with events")
t = np.linspace(0, 500, 10001)
sol = solve_ivp(rossler_sys, [t[0], t[-1]], sol_trans.y[:, -1], t_eval=t,
                events=x0event, dense_output=True,
                args=(a, b, c), method='Radau', jac=rossler_jac)

# Compute the state at each event.
w_events = sol.sol(sol.t_events[0])

plt.figure(1)
plt.title('Rössler phase space, (x, y) projection')
plt.plot(sol.y[0], sol.y[1], 'k', linewidth=1, alpha=0.25)
plt.plot(np.zeros(w_events.shape[1]), w_events[1], 'r.', markersize=2.5, alpha=0.5)
plt.xlabel('x')
plt.ylabel('y')

plt.figure(2)
plt.title('y return map at x=0')
plt.plot(w_events[1, :-1], w_events[1, 1:], 'r.', alpha=0.5)
plt.xlabel('y[k]')
plt.ylabel('y[k+1]')

plt.show()
```

The script generates the following plots:

![Figure_1](https://user-images.githubusercontent.com/321463/56779628-d14b0780-67a9-11e9-809d-a4d411faca53.png)
![Figure_2](https://user-images.githubusercontent.com/321463/56779633-d60fbb80-67a9-11e9-83df-41faf651c6a1.png)

rgommers commented:
This is a quite non-invasive change, and the absence of `args` clearly hinders enough people that it seems like a good idea to make this change. Showing `fun(t, y, ..., args=args)` and `solve_ivp(fun=lambda t, y: fun(t, y, *args), ...)` next to each other may even be good to teach people about lambdas.

This still needs tests, but +1 for merging.

mdhaber commented on scipy/integrate/_ivp/ivp.py:
```diff
@@ -492,6 +500,9 @@ def solve_ivp(fun, t_span, y0, method='RK45', t_eval=None, dense_output=False,
     if events is not None:
+        if args is not None:
+            original_events = events
+            events = [lambda t, x: event(t, x, *args) for event in original_events]
```
I think there is going to be an issue here. I think you'll run into this unintuitive thing when you define lambdas in a loop:
https://stackoverflow.com/questions/19837486/python-lambda-in-a-loop
The post contains the simple fix.

WarrenWeckesser replied:
Thanks. I thought about the late binding closures *gotcha* when I wrote that, but after I tried a variation of my demonstration script that had two event functions and it worked as expected, I put off thinking more about it.  I can create a problem with some simple examples in an ipython shell, so now I have to figure out why my demo worked (or why I *thought* it worked).

WarrenWeckesser replied:
I figured out why my example with two event functions appeared to work.  The functions returned the same value, but they had different directions.   Because the direction information has already been separated from the functions at this point in the code, my demo worked, even though both events were actually using the same function (the second one).

mdhaber replied:
A different example would be better for the tests : )

WarrenWeckesser replied:
Fixed.

mdhaber commented:
Great start! I'd be happy to take a final look after tests (I'd suggest including one with multiple events) and doc updates.

WarrenWeckesser commented:
I simplified the changes a bit, fixed the late binding closure problem, and added a test.

I still need to update the docstring to include a description of the `args` parameter.

mdhaber commented:
Could steal it from `minimize`:

args : tuple, optional
Extra arguments passed to the objective function and its derivatives (fun, jac and hess functions).

In this case I guess 'hess' should be 'event'. 
Or if you write something more detailed, maybe I'll copy that into `minimize`.

ilayn commented:
Very nice. I know I'm a bit late but do you think it is a bit better to name it as `fun_args` since first parameter is called `fun`? It is not a blocker but considering the standard `*args, *kwargs` tradition in Python it sounds like extra args to `solve_ivp`

WarrenWeckesser commented:
@ilayn, we already use `args` in many places: `integrate.odeint`, `integrate.quad`, `optimize.minimize`, `optimize.root`, `optimize.fmin`, etc.  I think it is better to maintain a consistent API and use `args` in `solve_ivp`.

ilayn commented:
Ah didn't know that. Nevermind then.

WarrenWeckesser commented:
Test added, docs updated, ready for review.

mdhaber commented on scipy/integrate/_ivp/ivp.py:
```diff
@@ -443,6 +454,32 @@ def solve_ivp(fun, t_span, y0, method='RK45', t_eval=None, dense_output=False,
+    >>> plt.legend(['x', 'y'], shadow=True)
+    >>> plt.title('Lotka-Volterra System')
+    >>> plt.show()
+
```
This all renders well (https://13037-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp).

mdhaber commented on scipy/integrate/_ivp/ivp.py:
```diff
@@ -451,6 +488,14 @@ def solve_ivp(fun, t_span, y0, method='RK45', t_eval=None, dense_output=False,
+    if args is not None:
+        # Wrap the user's fun (and jac, if given) in lambdas to hide the
+        # additional parameters.
+        fun = lambda t, x, fun=fun: fun(t, x, *args)
```
Curious - I know `event=event` is needed below, but is `fun=fun`/`jac=jac` needed here or just a precaution?

WarrenWeckesser replied:
`fun` is redefined by that statement.  If I didn't pass in `fun` like that, the original definition would be lost, and when the lambda was eventually called, it would then attempt to call itself, resulting in an error.

In the first version of the pull request, I had

        original_fun = fun
        fun = lambda t, x: original_fun(t, x, *args)

which is an alternative way to ensure that the lambda expression calls the original function.

WarrenWeckesser replied:
I noticed that `jac=jac` is not needed in the wrapper of `jac`, because the only `jac` defined is in the original `jac` in scope outside the lambda expression, so I removed it.

mdhaber commented on scipy/integrate/_ivp/ivp.py:
```diff
@@ -492,6 +537,10 @@ def solve_ivp(fun, t_span, y0, method='RK45', t_eval=None, dense_output=False,
 
     if events is not None:
+        if args is not None:
+            # Wrap user functions in lambdas to hide the additional parameters.
```
Would be good to explain why `event=event:` is needed.

WarrenWeckesser replied:
I added a comment (here and up by the `fun` lambda wrapper about using the keyword argument to keep the original function in scope.

mdhaber commented on scipy/integrate/tests/test_ivp.py:
```diff
@@ -791,3 +791,86 @@ def structure(n):
+        return [-omega*y, omega*x, k*z*(1 - z)]
+
+    def sys3_jac(t, w, omega, k, zfinal):
+        print(t)
```
Not intentional, right?

WarrenWeckesser replied:
Oops.  That was meant to be temporary.  I used it to verify that the Jacobian function was being called a few times.  It will be gone in the next update.

WarrenWeckesser replied:
Fixed.

mdhaber commented on scipy/integrate/tests/test_ivp.py:
```diff
@@ -791,3 +791,86 @@ def structure(n):
+    # sys3 is actually two decoupled systems. (x, y) form a
+    # linear oscillator, while z is a nonlinear first order
+    # system with equilibria at z=0 and z=1.  If k > 0, z=1
+    # is stable and z=0 is ubstable.
```
"unstable" typo

WarrenWeckesser replied:
Fixed.

mdhaber commented on scipy/integrate/tests/test_ivp.py:
```diff
@@ -791,3 +791,86 @@ def structure(n):
+    def sys3_jac(t, w, omega, k, zfinal):
+        print(t)
+        x, y, z = w
+        J = np.array([[0, -omega, 0],
```
Jacobian looks right.

mdhaber commented on scipy/integrate/tests/test_ivp.py:
```diff
@@ -791,3 +791,86 @@ def structure(n):
+    zfinal = 0.99
+    # Find z0 such that when z(0) = z0, z(tfinal) = zfinal.
+    # The condition z(tfinal) = zfinal is the terminal event.
+    z0 = np.exp(-k*tfinal)/((1 - zfinal)/zfinal + np.exp(-k*tfinal))
```
I'll trust you on this : ) Is there a reference?

WarrenWeckesser replied:
I'm not sure how to cite the scribbled calculations on the scrap of paper next to my computer. 😄

mdhaber replied:
I was _this_ close to creating a Wikipedia page about the "Weckesser Oscillator" : )

mdhaber replied:
I was _this_ close to creating a Wikipedia page about the Weckesser oscillator : )

mdhaber commented on scipy/integrate/_ivp/ivp.py:
```diff
@@ -443,6 +454,32 @@ def solve_ivp(fun, t_span, y0, method='RK45', t_eval=None, dense_output=False,
+    We pass in the parameter values a=1.5, b=1, c=3 and d=1 with the `args`
+    argument.
+
+    >>> sol = solve_ivp(lotkavolterra, [0, 15], [10, 5], args=(1.5, 1, 3, 1),
```
Would be nice to show the lambda version next to this as @rgommers suggested.

WarrenWeckesser replied:
I don't think we need that in the `solve_ivp` docstring.  I interpreted Ralf's suggestion as something that could be shown in a class when teaching lambda expressions.

WarrenWeckesser commented:
@mdhaber, thanks for the review.  I've responded to all your comments, and pushed an update with fixes.  (Tests are running now.)

mdhaber commented:
@ilayn what do you think? If you're OK with it, I think it's ready.
Commits look clean, so we can keep them separate?

WarrenWeckesser commented:
> Commits look clean, so we can keep them separate?

My opinion: I'm very much in favor of cleaning up commits when there is a lot of churn (e.g. fixing typos, bugs and other problems that were found in previous commits in the same pull request--we don't need that churn in the commit history).  In this case, there are just two commits: one that adds the new feature, and one that does some unrelated clean up in the docstring.  This is a case where it makes sense to keep them distinct.

mdhaber commented:
@WarrenWeckesser Recently one of my PRs was merged with all sorts of messy commits, and immediately after that I squash-merged someone else's nice, clean commits. Learned my lesson, so I thought I'd be very careful by asking : ) Based on what I know now, I agree that it makes perfect sense to keep them separate.

mdhaber commented:
Closes #8352

WarrenWeckesser commented:
@TejasAvinashShetty, in general, `args` will be a tuple containing the additional arguments.  If you have one argument, `omega_vec`, that happens to be an array, you pass it to the `args` parameter in a tuple with length 1,  e.g. `args=(omega_vec,)`.
